Restructure project

ilija-lazoroski opened the pull request.

ilija-lazoroski committed:
Improve README.md slightly

ilija-lazoroski committed:
Restructure the agent_plugin_builder

This way we can have a package with same name which exposes a
single method to build the plugin providing just a path

ilija-lazoroski committed:
Add build_plugin.py at root that can be used with cli argument

ilija-lazoroski committed:
Furhter improve README.md

ilija-lazoroski committed:
Implement __main__ for agent_plugin_builder

ilija-lazoroski committed:
Add needed files from template

ilija-lazoroski committed:
Use `build` and `dist` for resulting artifacts

cakekoa commented on .gitignore:
```diff
@@ -1,2 +1,160 @@
+*.mo
+*.pot
+
+# Django stuff:
```
Why not delete the stuff for tools that we aren't using?

mssalvatore replied:
It's from the general Python template supplied by GitHub. It shouldn't hurt anything to just leave it.

cakekoa commented on README.md:
```diff
@@ -1,8 +1,41 @@
+
+### Using Poetry
+
+You can also run it with `poetry run build_agent_plguin <PLUGIN_PATH>`.
```
Does it need to be installed in order to use `poetry run ...`? If not, then you should mention it. Perhaps say "alternatively, one may use Agent Plugin Builder without installing it by invoking `poetry run ...`"

cakekoa commented on README.md:
```diff
@@ -1,8 +1,41 @@
+
+## Resulting artifact
+
+All build artifacts used to generate the plugin archive will be in `build` directory
```
Relative to the plugin, or this repo?

cakekoa commented on README.md:
```diff
@@ -1,8 +1,41 @@
+## Resulting artifact
+
+All build artifacts used to generate the plugin archive will be in `build` directory
+and the Anget Plugin archive will be `dist` in the same directory when the
```
```suggestion
and the Agent Plugin archive will be `dist` in the same directory when the
```

cakekoa commented on README.md:
````diff
@@ -1,8 +1,41 @@
+1. Run the following commands to install the necessary prerequisites:
+
+    ```
+    $ pip install poetry pre-commit
````
Would it work just as well to make pre-commit a dev dependency? Or is that not kosher?

mssalvatore replied:
It's not something the project actually needs in order to run.  In other words, it's not required for the functional code, tests, or build. In my mind, it shouldn't go in the pyproject.toml. But I can't make a very strong argument for this.

cakekoa commented on agent_plugin_builder/__init__.py:
It may still be useful to export utility functions from this package, if we want to allow folks to create a custom build script

cakekoa commented on agent_plugin_builder/build_plugin.py:
```diff
@@ -28,17 +26,21 @@ def main():
-        prefix=f"build_{agent_plugin_manifest.plugin_type}_{agent_plugin_manifest.name}_"
-    )
-    build_dir = Path(tmp_build_dir)
+    plugin_build_dir = Path.cwd() / "build"
```
Can we make the build location customizable? Perhaps add an argument? Be wary of using a default argument

cakekoa commented on agent_plugin_builder/compare_package_lists.py:
```diff
@@ -14,6 +14,6 @@ def take(n: int, iterable) -> list[str]:
 
 
-def all_equal(packages: set[str]) -> bool:
+def all_equal(packages: list[set[str]]) -> bool:
```
`list` is too restrictive here. Use `Iterable`. Note that this is an [itertools recipe](https://docs.python.org/3/library/itertools.html#itertools-recipes). Since the groupby key is `None`, it should work for any type that can be compared via `==`, which AFAIK all objects have, so we could use `Iterable[Any]` here.

ilija-lazoroski committed:
Fix small notes in README

ilija-lazoroski committed:
Fix type hints in compare_package_lists

ilija-lazoroski committed:
Expose certain methods in __init__

ilija-lazoroski committed:
Add optional build and dist arguments

ilija-lazoroski committed:
Fix name inconsistency in build_plugin

ilija-lazoroski committed:
Lower plugin type in dist archive

ilija-lazoroski committed:
Define const for plugin specific names

Maybe we should configure this somehow

ilija-lazoroski committed:
Add configured logging to the scripts

cakekoa committed:
Fix default DIST_DIR_PATH value

cakekoa committed:
Remove default path arguments

cakekoa committed:
Fix spelling in plugin build script

cakekoa committed:
Simplify console logs

cakekoa committed:
Pipe subprocess stdout/stderr to the logger

ilija-lazoroski committed:
Move main method to different file

ilija-lazoroski committed:
Add docstrings to all methods

ilija-lazoroski committed:
Fix type hinting across project

ilija-lazoroski committed:
Clear the build directory if it exists

This will ensure that we won't have conflicts if there are plugin
specific code already there.

ilija-lazoroski committed:
Change '_' to '-' in resulting plugin archive

ilija-lazoroski committed:
Add exception handling when copy/removing build directory

ilija-lazoroski committed:
Fix README

ilija-lazoroski committed:
Compress source archive with separate files

cakekoa commented on agent_plugin_builder/build_plugin.py:
```diff
@@ -188,7 +188,8 @@ def create_source_archive(build_dir_path: Path):
     with tarfile.open(str(source_archive), "w:gz") as tar:
-        tar.add(source_build_dir_path, arcname=SOURCE, filter=_source_archive_filter)
+        for item in source_build_dir_path.iterdir():
+            tar.add(item, arcname=item.name, filter=_source_archive_filter)
```
As an alternative of iterating, specifying the source path and setting `arcname="/"` in `tar.add()` appears to work as well

mssalvatore commented on .cruft.json:
```diff
@@ -0,0 +1,37 @@
+      "python_major_version": "3",
+      "python_minor_version": "11",
+      "support_type_checking": "y",
+      "hook_shellcheck": "n",
```
This should be enabled.

mssalvatore commented on .cruft.json:
```diff
@@ -0,0 +1,37 @@
+{
+  "template": "/home/mcus/repos/cookiecutter_agent_plugin_builder/../infection-monkey-cookiecutter",
```
This should point to the repository in github, otherwise no one but @ilija-lazoroski can ever update cruft.

cakekoa committed:
Use cookiecutter template from github

cakekoa committed:
Update template, adding shellcheck

mssalvatore commented on agent_plugin_builder/vendor_dirs.py:
```diff
@@ -59,70 +66,140 @@ def check_if_common_vendor_dir_possible(build_dir: Path, uid, gid) -> bool:
+        "chmod 755 /plugin/src/vendor-linux",
     ]
 
     full_command = "/bin/bash -c '" + " && ".join(commands) + "'"
```
We should be very careful about concatenating shell commands so that we avoid any risk of OS command injection. Ideally we wouldn't be dynamically building shell commands.

cakekoa committed:
Allow build directory to be cleared without affecting log file

cakekoa committed:
Write unhandled exception to the log file

cakekoa committed:
Escape shell command inputs

cakekoa committed:
Extract function _run_container_with_plugin_dir

cakekoa committed:
Fix __main__ pointing to invalid main function

ilija-lazoroski commented on agent_plugin_builder/vendor_dirs.py:
```diff
@@ -110,6 +100,16 @@ def _build_bash_command(command: str) -> str:
     return f"/bin/bash -c {quote(command)}"
 
 
+def _run_container_with_plugin_dir(image: str, command: str, plugin_dir: Path):
```
Missing return type

cakekoa replied:
The return type is `Container`

ilija-lazoroski replied:
I think not https://docker-py.readthedocs.io/en/stable/containers.html

cakekoa replied:
Hmm, you are right. The names of variables we're using to capture the result of this call are misleading. I think we should either:
1. Change the variable names, or
2. Output a container, and use the `wait()` and `logs()` methods to do what we need

ilija-lazoroski commented on agent_plugin_builder/vendor_dirs.py:
```diff
@@ -110,6 +100,16 @@ def _build_bash_command(command: str) -> str:
 
+def _run_container_with_plugin_dir(image: str, command: str, plugin_dir: Path):
+    """
+    Run a container with the plugin directory mounted.
```
Missing `:param`s

ilija-lazoroski committed:
Fix types in vendor directories

mssalvatore commented on agent_plugin_builder/agent_plugin_builder.py:
```diff
@@ -0,0 +1,67 @@
+        "DEBUG=5. Default(0 or more than 5 occurrences)  level: INFO.",
+    )
+    args = parser.parse_args()
+    _temp_log_dir = Path(mkdtemp())
+    _setup_logging(_temp_log_dir, args.verbosity)
+    _log_arguments(args)
+    try:
+        build_agent_plugin(args.plugin_path, args.build_dir_path, args.dist_dir_path)
+    except Exception as e:
+        logger.error(f"Error building plugin: {e}", exc_info=True)
+
+    logger.info(f"Copying log file to {args.build_dir_path}")
+    import shutil
+
+    shutil.copy(_temp_log_dir / AGENT_PLUGIN_BUILDER_LOG_FILENAME, args.build_dir_path)
```
I'd recommend we not write anything to /tmp. Anything we need to output before the log file is set up can be output to STDERR. Anything else can be output after the log file is created.

mssalvatore commented on agent_plugin_builder/build_options.py:
```diff
@@ -35,7 +38,25 @@ class AgentPluginBuildOptions(InfectionMonkeyBaseModel):
+    if not build_config_file_path.exists():
+        build_config_file_path = plugin_path / "build.yml"
+
+    if not build_config_file_path.exists():
+        logger.info(
+            "Build options not found, using default dependency method: "
+            f"{PlatformDependencyPackagingMethod.SEPARATE}"
+        )
+        return AgentPluginBuildOptions(
+            platform_dependencies=PlatformDependencyPackagingMethod.SEPARATE
+        )
```
NIT: To avoid unnecessary revisions if we ever add new options, we could be more vague here and just say "Using default options". We can also then just do:

```suggestion
    if not build_config_file_path.exists():
        logger.info(
            "Build options not found, using default dependency method: "
            f"{PlatformDependencyPackagingMethod.SEPARATE}"
        )
        return AgentPluginBuildOptions()
```

mssalvatore commented on agent_plugin_builder/build_plugin.py:
```diff
@@ -0,0 +1,252 @@
+    :param dist_dir_path: Path to the dist directory.
+        If the directory does not exist, it will be created
+    """
+    import shutil
```
Is there a reason this isn't at the top of the file?

mssalvatore commented on agent_plugin_builder/build_plugin.py:
```diff
@@ -0,0 +1,252 @@
+    dist_dir_path: Path,
+):
+    """
+    Build the agent plugin by copying the plugin code to the build directory,
+    generating the Agent Plugin archive.
```
```suggestion
    Build the agent plugin by copying the plugin code to the build directory
    and generating the Agent Plugin archive.
```

mssalvatore commented on agent_plugin_builder/build_plugin.py:
```diff
@@ -0,0 +1,252 @@
+    Build the agent plugin by copying the plugin code to the build directory,
+    generating the Agent Plugin archive.
+
+    :param plugin_path: Path to the plugin code.
+        If the directory does not exist, an error will be raised
```
```suggestion
    :param plugin_path: Path to the plugin code.
```

Instead, add ":raises FileNotFoundError: If blah blah".

mssalvatore commented on agent_plugin_builder/build_plugin.py:
```diff
@@ -0,0 +1,252 @@
+
+    if not build_dir_path.exists():
+        logger.info(f"Creating build directory: {build_dir_path}")
+        build_dir_path.mkdir(exist_ok=True)
```
Do we need to do this if `shutil.copytree()` will handle it down below?

mssalvatore commented on agent_plugin_builder/build_plugin.py:
```diff
@@ -0,0 +1,252 @@
+    create_agent_plugin_archive(build_dir_path, agent_plugin_manifest, dist_dir_path)
+
+
+def get_agent_plugin_manifest(build_dir_path: Path) -> AgentPluginManifest:
+    """
+    Get the Agent Plugin manifest from the build directory.
+
+    :param build_dir_path: Path to the build directory.
+    """
```
This could raise some kind of yaml error, and/or a file not found error (or similar).

mssalvatore commented on agent_plugin_builder/build_options.py:
```diff
@@ -35,7 +38,25 @@ class AgentPluginBuildOptions(InfectionMonkeyBaseModel):
+            f"{PlatformDependencyPackagingMethod.SEPARATE}"
+        )
+        return AgentPluginBuildOptions(
+            platform_dependencies=PlatformDependencyPackagingMethod.SEPARATE
```
I feel like `AUTODETECT` is a better default and the model should be modified accordingly.

cakekoa replied:
I used `SEPARATE` because it is the most reliable method. `AUTODETECT` is useful but I think it has a chance of incorrectly choosing to go with a common vendor dir in some cases (perhaps with a source dependency).

mssalvatore commented on agent_plugin_builder/build_plugin.py:
```diff
@@ -0,0 +1,252 @@
+    dependency_method: PlatformDependencyPackagingMethod,
+):
+    """
+    Generate the vendor directories for the plugin. If the dependency method is not chosen
+    or if the plugin supports multiple operating systems, the method will try to generate
+    common vendor directories. If common vendor directory are not possible, it will generate
+    separate vendor directories for each supported operating system.
```
"If the dependency method is not chosen"

Since this can't be `None`, a method is always chosen.

mssalvatore commented on agent_plugin_builder/vendor_dirs.py:
```diff
@@ -1,54 +1,81 @@
+        WINDOWS_BUILD_PACKAGE_LIST_COMMANDS.format(uid=quote(str(uid)), gid=quote(str(gid)))
     )
+    windows_container = _run_container_with_plugin_dir(PLUGIN_BUILDER_IMAGE, command, build_dir)
+    _log_container_output(windows_container, "Windows Dry Run requirements, ")
```
It would probably be better if the "prefix" was a string that didn't assume some kind of formatting. In other words, `_log_container_output()` should be responsible for handling commas, spacing, whatever.

mssalvatore commented on agent_plugin_builder/vendor_dirs.py:
```diff
@@ -1,54 +1,81 @@
+logger = logging.getLogger(__name__)
+
+
+AGENT_PLUGIN_IMAGE: Final = "infectionmonkey/agent-builder:latest"
+PLUGIN_BUILDER_IMAGE: Final = "infectionmonkey/plugin-builder:latest"
```
It might be better if theses were named "LINUX_PLUGIN_BUILDER_IMAGE" and "WINDOWS_PLUGIN_BUILDER_IMAGE". Since that's how they're used, the names would better capture their intent. Plus, it may not always be the case that Linux plugins are built with the agent-builder image.

mssalvatore commented on agent_plugin_builder/vendor_dirs.py:
```diff
@@ -60,71 +87,135 @@ def check_if_common_vendor_dir_possible(build_dir: Path, uid, gid) -> bool:
         command = ["poetry", "export", "-f", "requirements.txt", "-o", "requirements.txt"]
-        subprocess.check_call(command, cwd=str(build_dir))
+        process = subprocess.Popen(
+            command, cwd=str(build_dir), stdout=subprocess.PIPE, stderr=subprocess.STDOUT
+        )
+        with process.stdout as stdout:  # type: ignore [union-attr]
+            for line in iter(stdout.readline, b""):
+                logger.debug(line)
+
+        return_code = process.wait()
+        if return_code != 0:
+            raise subprocess.CalledProcessError(return_code, command)
     elif (build_dir / "Pipfile.lock").exists():
         command = ["pipenv", "requirements"]
         with (build_dir / "requirements.txt").open("w") as f:
-            subprocess.check_call(command, cwd=str(build_dir), stdout=f)
+            process = subprocess.Popen(
+                command, cwd=str(build_dir), stdout=f, stderr=subprocess.PIPE
+            )
+            with process.stderr as stderr:  # type: ignore [union-attr]
+                for line in iter(stderr.readline, b""):
+                    logger.debug(line)
+
+        return_code = process.wait()
+        if return_code != 0:
+            raise subprocess.CalledProcessError(return_code, command)
+
```
We could, perhaps, reduce duplication, here. More importantly, do we need to support pipenv? Since our cookiecutter sets up poetry and doesn't have any option for pipenv, I don't think we really need to support it. I'd prefer we only need to support one (poetry).

cakekoa replied:
We don't _need_ to. I'm fine getting rid of it

cakekoa replied:
The argument in favor of supporting pipenv is that all of our current plugins use it. However, if we only use this project to build plugins that we've extracted to their own repositories, and thus use poetry, then we needn't support pipenv

cakekoa replied:
That said, I removed it. One could get it working with pipenv by pre-generating the `requirements.txt`

mssalvatore commented on agent_plugin_builder/compare_package_lists.py:
```diff
@@ -9,11 +10,11 @@ def load_package_names(file_path: Path) -> set[str]:
 
 
-def all_equal(packages: set[str]) -> bool:
+def all_equal(packages: Iterable[Any]) -> bool:
```
I don't fully understand what this is doing, so I'm not saying it definitely won't work as intended. However, the [itertools documentation](https://docs.python.org/3/library/itertools.html#itertools.groupby) says the inputs to `groupby()` must be sorted. But, what gets passed in from `check_if_common_vendor_dir_possible()` is the return values from `load_package_names()`. `load_package_names()` returns a set, and sets in Python are not ordered.

Since `linux_packages` and `windows_packages` are already sets, can we just say, `if linux_packages == windows packages`?

cakekoa replied:
Good catch

mssalvatore commented on agent_plugin_builder/vendor_dirs.py:
```diff
@@ -60,71 +87,135 @@ def check_if_common_vendor_dir_possible(build_dir: Path, uid, gid) -> bool:
     windows_packages_path.unlink()
 
     response = all_equal([linux_packages, windows_packages])
+    if response:
+        logger.info("Common vendor directory is possible")
```
```suggestion
    if linux_packages == windows_packages:
        logger.info("Common vendor directory is possible")
```

See https://github.com/guardicode/agent-plugin-builder/pull/1/files#r1657406678 for an explanation.

mssalvatore commented on agent_plugin_builder/build_plugin.py:
```diff
@@ -0,0 +1,252 @@
+    return file_info
+
+
+def create_plugin_archive(
```
In my experience, it's common for _all_ of the build artifacts to be created in the `build/` directory and the artifacts for distribution are _copied_ to the `dist/` directory after the build completes.

cakekoa committed:
Simplify platform package list comparison

cakekoa committed:
Drop support for pipenv

cakekoa committed:
Remove pipenv dependency

cakekoa committed:
Rename image variables to match their purpose

cakekoa committed:
Update _log_container_output to handle formatting

cakekoa committed:
Update generate_vendor_directories description

cakekoa committed:
Add raises to get_agent_plugin_manifest docstring

cakekoa committed:
Rely on copytree() to create the build dir

cakekoa committed:
Simplify log message for default build options

cakekoa committed:
Update docstring for build_agent_plugin

cakekoa committed:
Only log to file after build directory is created

cakekoa committed:
Extract function _get_plugin_archive_name

cakekoa committed:
Create plugin in build dir, and copy to dist dir

ilija-lazoroski commented on agent_plugin_builder/build_plugin.py:
```diff
@@ -0,0 +1,269 @@
+    plugin_path: Path,
+    build_dir_path: Path,
+    dist_dir_path: Path,
+    on_build_dir_created: Callable[[Path], None] | None = None,
```
This complicates stuff. We don't need a callable for the log file. The `build_agent_plugin` will eventually be exposed and it needs to be simple.

cakekoa replied:
It is an _optional_ argument -- one need not provide it

ilija-lazoroski commented on agent_plugin_builder/vendor_dirs.py:
```diff
@@ -59,72 +85,129 @@ def check_if_common_vendor_dir_possible(build_dir: Path, uid, gid) -> bool:
     linux_packages_path.unlink()
     windows_packages_path.unlink()
```
We probably don't need to unlink these as we keep them in the build directory and not cleaning anything else.

cakekoa replied:
Done. It could be useful for debugging. I also changed the filenames to be more descriptive

cakekoa committed:
Leave generated platform package lists in the build dir

cakekoa committed:
Move "copy to dist dir" out of create_plugin_archive

cakekoa committed:
Update create_source_archive to return path to created archive

cakekoa committed:
Document function return values

cakekoa committed:
Add `source-dirname` as argument

As you can choose the name of your plugin repository
we are providing an argument so you can choose the plugin source
code directory

cakekoa committed:
Simplify run instructions

cakekoa committed:
Simplify docker commands

cakekoa committed:
Add source_dir as a build option

cakekoa committed:
Pass command-line options as build options
